<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>3 Correlation | Notes of R for Bioinformatics</title>
  <meta name="description" content="3 Correlation | Notes of R for Bioinformatics">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="3 Correlation | Notes of R for Bioinformatics" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3 Correlation | Notes of R for Bioinformatics" />
  
  
  

<meta name="author" content="Shaojun Xie">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="plot-in-r.html">
<link rel="next" href="standard-deviation-vs-standard-error.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="introduction-of-r.html"><a href="introduction-of-r.html"><i class="fa fa-check"></i><b>1</b> Introduction of R</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction-of-r.html"><a href="introduction-of-r.html#setup"><i class="fa fa-check"></i><b>1.1</b> Setup</a></li>
<li class="chapter" data-level="1.2" data-path="introduction-of-r.html"><a href="introduction-of-r.html#basic-systax"><i class="fa fa-check"></i><b>1.2</b> Basic systax</a><ul>
<li class="chapter" data-level="1.2.1" data-path="introduction-of-r.html"><a href="introduction-of-r.html#install-packages"><i class="fa fa-check"></i><b>1.2.1</b> Install packages</a></li>
<li class="chapter" data-level="1.2.2" data-path="introduction-of-r.html"><a href="introduction-of-r.html#install"><i class="fa fa-check"></i><b>1.2.2</b> Install</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="introduction-of-r.html"><a href="introduction-of-r.html#data-type"><i class="fa fa-check"></i><b>1.3</b> Data type</a></li>
<li class="chapter" data-level="1.4" data-path="introduction-of-r.html"><a href="introduction-of-r.html#data-structures"><i class="fa fa-check"></i><b>1.4</b> Data structures</a></li>
<li class="chapter" data-level="1.5" data-path="introduction-of-r.html"><a href="introduction-of-r.html#read-from-and-write-into-files"><i class="fa fa-check"></i><b>1.5</b> Read from and write into files</a><ul>
<li class="chapter" data-level="1.5.1" data-path="introduction-of-r.html"><a href="introduction-of-r.html#read-random-rows-from-a-huge-file"><i class="fa fa-check"></i><b>1.5.1</b> Read Random Rows from A Huge File</a></li>
<li class="chapter" data-level="1.5.2" data-path="introduction-of-r.html"><a href="introduction-of-r.html#reference"><i class="fa fa-check"></i><b>1.5.2</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="1.6" data-path="introduction-of-r.html"><a href="introduction-of-r.html#functions"><i class="fa fa-check"></i><b>1.6</b> Functions</a></li>
<li class="chapter" data-level="1.7" data-path="introduction-of-r.html"><a href="introduction-of-r.html#control-structure"><i class="fa fa-check"></i><b>1.7</b> Control structure</a></li>
<li class="chapter" data-level="1.8" data-path="introduction-of-r.html"><a href="introduction-of-r.html#reference-1"><i class="fa fa-check"></i><b>1.8</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="plot-in-r.html"><a href="plot-in-r.html"><i class="fa fa-check"></i><b>2</b> Plot in R</a><ul>
<li class="chapter" data-level="2.1" data-path="plot-in-r.html"><a href="plot-in-r.html#read-the-data"><i class="fa fa-check"></i><b>2.1</b> Read the data</a></li>
<li class="chapter" data-level="2.2" data-path="plot-in-r.html"><a href="plot-in-r.html#line-plot"><i class="fa fa-check"></i><b>2.2</b> Line plot</a></li>
<li class="chapter" data-level="2.3" data-path="plot-in-r.html"><a href="plot-in-r.html#basic-line-plot"><i class="fa fa-check"></i><b>2.3</b> Basic line plot</a></li>
<li class="chapter" data-level="2.4" data-path="plot-in-r.html"><a href="plot-in-r.html#color-the-dots-based-on-the-sex-information"><i class="fa fa-check"></i><b>2.4</b> Color the dots based on the <code>sex</code> information</a></li>
<li class="chapter" data-level="2.5" data-path="plot-in-r.html"><a href="plot-in-r.html#section"><i class="fa fa-check"></i><b>2.5</b> </a></li>
<li class="chapter" data-level="2.6" data-path="plot-in-r.html"><a href="plot-in-r.html#draw-a-circle-in-r"><i class="fa fa-check"></i><b>2.6</b> Draw a circle in R</a></li>
<li class="chapter" data-level="2.7" data-path="plot-in-r.html"><a href="plot-in-r.html#network"><i class="fa fa-check"></i><b>2.7</b> Network</a><ul>
<li class="chapter" data-level="2.7.1" data-path="plot-in-r.html"><a href="plot-in-r.html#section-1"><i class="fa fa-check"></i><b>2.7.1</b> </a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="correlation.html"><a href="correlation.html"><i class="fa fa-check"></i><b>3</b> Correlation</a><ul>
<li class="chapter" data-level="3.1" data-path="correlation.html"><a href="correlation.html#visulization-of-pair-wise-correlation-in-r"><i class="fa fa-check"></i><b>3.1</b> Visulization of pair-wise correlation in R</a></li>
<li class="chapter" data-level="3.2" data-path="correlation.html"><a href="correlation.html#correlation-and-p-values-of-all-combinations-of-all-rows-of-two-matrices"><i class="fa fa-check"></i><b>3.2</b> Correlation and p-values of all combinations of all rows of two matrices</a><ul>
<li class="chapter" data-level="3.2.1" data-path="correlation.html"><a href="correlation.html#prepare-the-data"><i class="fa fa-check"></i><b>3.2.1</b> Prepare the data</a></li>
<li class="chapter" data-level="3.2.2" data-path="correlation.html"><a href="correlation.html#another-way-to-do-this-is-to-use-corr.test"><i class="fa fa-check"></i><b>3.2.2</b> Another way to do this is to use <code>corr.test</code></a></li>
<li class="chapter" data-level="3.2.3" data-path="correlation.html"><a href="correlation.html#read-data"><i class="fa fa-check"></i><b>3.2.3</b> Read data</a></li>
<li class="chapter" data-level="3.2.4" data-path="correlation.html"><a href="correlation.html#calculate-the-correlation-advance"><i class="fa fa-check"></i><b>3.2.4</b> Calculate the correlation (Advance)</a></li>
<li class="chapter" data-level="3.2.5" data-path="correlation.html"><a href="correlation.html#reference-3"><i class="fa fa-check"></i><b>3.2.5</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="correlation.html"><a href="correlation.html#different-ways-to-visulize-correlation"><i class="fa fa-check"></i><b>3.3</b> Different ways to visulize correlation</a><ul>
<li class="chapter" data-level="3.3.1" data-path="correlation.html"><a href="correlation.html#display-correlation-using-chart.correlation-in-performanceanalytics"><i class="fa fa-check"></i><b>3.3.1</b> Display correlation using <code>chart.Correlation</code> in <code>PerformanceAnalytics</code></a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="correlation.html"><a href="correlation.html#reference-4"><i class="fa fa-check"></i><b>3.4</b> Reference</a></li>
</ul></li>
<li class="part"><span><b>II Statistics</b></span></li>
<li class="chapter" data-level="4" data-path="standard-deviation-vs-standard-error.html"><a href="standard-deviation-vs-standard-error.html"><i class="fa fa-check"></i><b>4</b> Standard deviation vs Standard error</a><ul>
<li class="chapter" data-level="4.1" data-path="standard-deviation-vs-standard-error.html"><a href="standard-deviation-vs-standard-error.html#reference-5"><i class="fa fa-check"></i><b>4.1</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="t-test.html"><a href="t-test.html"><i class="fa fa-check"></i><b>5</b> T test</a><ul>
<li class="chapter" data-level="5.1" data-path="t-test.html"><a href="t-test.html#calculate-t-statistic-step-by-step-in-r"><i class="fa fa-check"></i><b>5.1</b> Calculate t-statistic step by step in R</a><ul>
<li class="chapter" data-level="5.1.1" data-path="t-test.html"><a href="t-test.html#section-2"><i class="fa fa-check"></i><b>5.1.1</b> </a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="t-test.html"><a href="t-test.html#calculate-the-standard-deviation-in-r"><i class="fa fa-check"></i><b>5.2</b> Calculate the standard deviation in R</a></li>
<li class="chapter" data-level="5.3" data-path="t-test.html"><a href="t-test.html#reference-6"><i class="fa fa-check"></i><b>5.3</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="linear-model.html"><a href="linear-model.html"><i class="fa fa-check"></i><b>6</b> Linear model</a><ul>
<li class="chapter" data-level="6.1" data-path="linear-model.html"><a href="linear-model.html#section-3"><i class="fa fa-check"></i><b>6.1</b> </a></li>
<li class="chapter" data-level="6.2" data-path="linear-model.html"><a href="linear-model.html#extract-male-data"><i class="fa fa-check"></i><b>6.2</b> Extract male data</a></li>
<li class="chapter" data-level="6.3" data-path="linear-model.html"><a href="linear-model.html#reference-7"><i class="fa fa-check"></i><b>6.3</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="mrg.html"><a href="mrg.html"><i class="fa fa-check"></i><b>7</b> MRG</a><ul>
<li class="chapter" data-level="7.1" data-path="mrg.html"><a href="mrg.html#read-the-data-1"><i class="fa fa-check"></i><b>7.1</b> Read the data</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="multiple-linear-regression.html"><a href="multiple-linear-regression.html"><i class="fa fa-check"></i><b>8</b> Multiple linear regression</a><ul>
<li class="chapter" data-level="8.1" data-path="multiple-linear-regression.html"><a href="multiple-linear-regression.html#section-4"><i class="fa fa-check"></i><b>8.1</b> </a></li>
<li class="chapter" data-level="8.2" data-path="multiple-linear-regression.html"><a href="multiple-linear-regression.html#section-5"><i class="fa fa-check"></i><b>8.2</b> </a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="bayesian-newwork-in-r.html"><a href="bayesian-newwork-in-r.html"><i class="fa fa-check"></i><b>9</b> Bayesian newwork in R</a><ul>
<li class="chapter" data-level="9.1" data-path="bayesian-newwork-in-r.html"><a href="bayesian-newwork-in-r.html#practical-experiments"><i class="fa fa-check"></i><b>9.1</b> Practical experiments</a><ul>
<li class="chapter" data-level="9.1.1" data-path="bayesian-newwork-in-r.html"><a href="bayesian-newwork-in-r.html#analysis-of-the-magic-population-in-scutari-et-al.-genetics-2014"><i class="fa fa-check"></i><b>9.1.1</b> Analysis of the MAGIC population in Scutari et al., Genetics (2014)</a></li>
<li class="chapter" data-level="9.1.2" data-path="bayesian-newwork-in-r.html"><a href="bayesian-newwork-in-r.html#reference-8"><i class="fa fa-check"></i><b>9.1.2</b> Reference</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="the-rationale-for-using-negative-binomial-distribution-to-model-read-count-in-rna-seq-data.html"><a href="the-rationale-for-using-negative-binomial-distribution-to-model-read-count-in-rna-seq-data.html"><i class="fa fa-check"></i><b>10</b> The rationale for using negative binomial distribution to model read count in RNA-seq data</a><ul>
<li class="chapter" data-level="10.1" data-path="the-rationale-for-using-negative-binomial-distribution-to-model-read-count-in-rna-seq-data.html"><a href="the-rationale-for-using-negative-binomial-distribution-to-model-read-count-in-rna-seq-data.html#negative-binomial-in-r"><i class="fa fa-check"></i><b>10.1</b> Negative binomial in R</a><ul>
<li class="chapter" data-level="10.1.1" data-path="the-rationale-for-using-negative-binomial-distribution-to-model-read-count-in-rna-seq-data.html"><a href="the-rationale-for-using-negative-binomial-distribution-to-model-read-count-in-rna-seq-data.html#real-data"><i class="fa fa-check"></i><b>10.1.1</b> Real data</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="the-rationale-for-using-negative-binomial-distribution-to-model-read-count-in-rna-seq-data.html"><a href="the-rationale-for-using-negative-binomial-distribution-to-model-read-count-in-rna-seq-data.html#reference-9"><i class="fa fa-check"></i><b>10.2</b> Reference</a></li>
<li class="chapter" data-level="10.3" data-path="the-rationale-for-using-negative-binomial-distribution-to-model-read-count-in-rna-seq-data.html"><a href="the-rationale-for-using-negative-binomial-distribution-to-model-read-count-in-rna-seq-data.html#batch-effect"><i class="fa fa-check"></i><b>10.3</b> Batch effect</a></li>
<li class="chapter" data-level="10.4" data-path="the-rationale-for-using-negative-binomial-distribution-to-model-read-count-in-rna-seq-data.html"><a href="the-rationale-for-using-negative-binomial-distribution-to-model-read-count-in-rna-seq-data.html#reference-10"><i class="fa fa-check"></i><b>10.4</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="multiple-test-corretion.html"><a href="multiple-test-corretion.html"><i class="fa fa-check"></i><b>11</b> Multiple test corretion</a></li>
<li class="chapter" data-level="12" data-path="working-with-missing-values.html"><a href="working-with-missing-values.html"><i class="fa fa-check"></i><b>12</b> Working with missing values</a><ul>
<li class="chapter" data-level="12.1" data-path="working-with-missing-values.html"><a href="working-with-missing-values.html#section-6"><i class="fa fa-check"></i><b>12.1</b> </a></li>
<li class="chapter" data-level="12.2" data-path="working-with-missing-values.html"><a href="working-with-missing-values.html#summaryize-the-missing-values"><i class="fa fa-check"></i><b>12.2</b> Summaryize the missing values</a><ul>
<li class="chapter" data-level="12.2.1" data-path="working-with-missing-values.html"><a href="working-with-missing-values.html#summarize-the-missing-values-using-mice-package"><i class="fa fa-check"></i><b>12.2.1</b> Summarize the missing values using <code>mice</code> package</a></li>
<li class="chapter" data-level="12.2.2" data-path="working-with-missing-values.html"><a href="working-with-missing-values.html#summarize-the-missing-values-using-vim-packages"><i class="fa fa-check"></i><b>12.2.2</b> Summarize the missing values using <code>VIM</code> packages</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="working-with-missing-values.html"><a href="working-with-missing-values.html#delete-columnsrows-with-more-that-x-missing"><i class="fa fa-check"></i><b>12.3</b> Delete columns/rows with more that x% missing</a></li>
<li class="chapter" data-level="12.4" data-path="working-with-missing-values.html"><a href="working-with-missing-values.html#imputing-the-data"><i class="fa fa-check"></i><b>12.4</b> Imputing the data</a><ul>
<li class="chapter" data-level="12.4.1" data-path="working-with-missing-values.html"><a href="working-with-missing-values.html#imputing-the-data-with-row-wise-mean"><i class="fa fa-check"></i><b>12.4.1</b> Imputing the data with row-wise mean</a></li>
<li class="chapter" data-level="12.4.2" data-path="working-with-missing-values.html"><a href="working-with-missing-values.html#imputing-the-data-with-row-wise-mean-using-mice"><i class="fa fa-check"></i><b>12.4.2</b> Imputing the data with row-wise mean using <code>mice</code></a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="working-with-missing-values.html"><a href="working-with-missing-values.html#reference-11"><i class="fa fa-check"></i><b>12.5</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>13</b> Data wrangling</a><ul>
<li class="chapter" data-level="13.1" data-path="data-wrangling.html"><a href="data-wrangling.html#section-7"><i class="fa fa-check"></i><b>13.1</b> </a></li>
<li class="chapter" data-level="13.2" data-path="data-wrangling.html"><a href="data-wrangling.html#joining-data-in-r-with-dplyr"><i class="fa fa-check"></i><b>13.2</b> Joining Data in R with dplyr</a><ul>
<li class="chapter" data-level="13.2.1" data-path="data-wrangling.html"><a href="data-wrangling.html#whats-covered"><i class="fa fa-check"></i><b>13.2.1</b> Whats Covered</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="data-wrangling.html"><a href="data-wrangling.html#reference-12"><i class="fa fa-check"></i><b>13.3</b> Reference</a></li>
</ul></li>
<li class="part"><span><b>III Machine learning</b></span></li>
<li class="chapter" data-level="14" data-path="pca.html"><a href="pca.html"><i class="fa fa-check"></i><b>14</b> PCA</a><ul>
<li class="chapter" data-level="14.1" data-path="pca.html"><a href="pca.html#read-data-1"><i class="fa fa-check"></i><b>14.1</b> Read data</a></li>
<li class="chapter" data-level="14.2" data-path="pca.html"><a href="pca.html#head-data"><i class="fa fa-check"></i><b>14.2</b> Head data</a></li>
<li class="chapter" data-level="14.3" data-path="pca.html"><a href="pca.html#summary-of-the-data"><i class="fa fa-check"></i><b>14.3</b> Summary of the data</a></li>
<li class="chapter" data-level="14.4" data-path="pca.html"><a href="pca.html#check-the-distribution-of-the-data"><i class="fa fa-check"></i><b>14.4</b> Check the distribution of the data</a></li>
<li class="chapter" data-level="14.5" data-path="pca.html"><a href="pca.html#eigendecomposition---computing-eigenvectors-and-eigenvalues"><i class="fa fa-check"></i><b>14.5</b> Eigendecomposition - Computing Eigenvectors and Eigenvalues</a><ul>
<li class="chapter" data-level="14.5.1" data-path="pca.html"><a href="pca.html#covariance-matrix"><i class="fa fa-check"></i><b>14.5.1</b> Covariance Matrix</a></li>
</ul></li>
<li class="chapter" data-level="14.6" data-path="pca.html"><a href="pca.html#pca-with-two-variables"><i class="fa fa-check"></i><b>14.6</b> PCA with two variables</a></li>
<li class="chapter" data-level="14.7" data-path="pca.html"><a href="pca.html#pca-on-wine-data"><i class="fa fa-check"></i><b>14.7</b> PCA on <code>wine</code> data</a><ul>
<li class="chapter" data-level="14.7.1" data-path="pca.html"><a href="pca.html#read-the-wine-data"><i class="fa fa-check"></i><b>14.7.1</b> Read the <code>wine</code> data</a></li>
<li class="chapter" data-level="14.7.2" data-path="pca.html"><a href="pca.html#description-of-each-column"><i class="fa fa-check"></i><b>14.7.2</b> Description of each column</a></li>
<li class="chapter" data-level="14.7.3" data-path="pca.html"><a href="pca.html#example"><i class="fa fa-check"></i><b>14.7.3</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="14.8" data-path="pca.html"><a href="pca.html#references"><i class="fa fa-check"></i><b>14.8</b> References</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="random-forest.html"><a href="random-forest.html"><i class="fa fa-check"></i><b>15</b> Random forest</a><ul>
<li class="chapter" data-level="15.1" data-path="random-forest.html"><a href="random-forest.html#practical-experiments-1"><i class="fa fa-check"></i><b>15.1</b> Practical experiments</a><ul>
<li class="chapter" data-level="15.1.1" data-path="random-forest.html"><a href="random-forest.html#random-forest-for-prediction-of-iris"><i class="fa fa-check"></i><b>15.1.1</b> Random forest for prediction of iris</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="qtl-analysis.html"><a href="qtl-analysis.html"><i class="fa fa-check"></i><b>16</b> QTL analysis</a><ul>
<li class="chapter" data-level="16.1" data-path="qtl-analysis.html"><a href="qtl-analysis.html#reference-13"><i class="fa fa-check"></i><b>16.1</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references-1.html"><a href="references-1.html"><i class="fa fa-check"></i>References</a></li>
<li class="chapter" data-level="17" data-path="doing-bayesian-statistics-in-r.html"><a href="doing-bayesian-statistics-in-r.html"><i class="fa fa-check"></i><b>17</b> Doing Bayesian statistics in R</a><ul>
<li class="chapter" data-level="17.1" data-path="doing-bayesian-statistics-in-r.html"><a href="doing-bayesian-statistics-in-r.html#reference-14"><i class="fa fa-check"></i><b>17.1</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="transformation-of-data.html"><a href="transformation-of-data.html"><i class="fa fa-check"></i><b>18</b> Transformation of data</a><ul>
<li class="chapter" data-level="18.1" data-path="transformation-of-data.html"><a href="transformation-of-data.html#log-transformation"><i class="fa fa-check"></i><b>18.1</b> Log transformation</a><ul>
<li class="chapter" data-level="18.1.1" data-path="transformation-of-data.html"><a href="transformation-of-data.html#the-rationale-of-using-log2-transformation"><i class="fa fa-check"></i><b>18.1.1</b> The rationale of using log2 transformation</a></li>
<li class="chapter" data-level="18.1.2" data-path="transformation-of-data.html"><a href="transformation-of-data.html#reference-15"><i class="fa fa-check"></i><b>18.1.2</b> Reference</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="goodness-of-fit-in-r.html"><a href="goodness-of-fit-in-r.html"><i class="fa fa-check"></i><b>19</b> Goodness of fit in R</a><ul>
<li class="chapter" data-level="19.1" data-path="goodness-of-fit-in-r.html"><a href="goodness-of-fit-in-r.html#reference-16"><i class="fa fa-check"></i><b>19.1</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="overdispersion.html"><a href="overdispersion.html"><i class="fa fa-check"></i><b>20</b> Overdispersion</a><ul>
<li class="chapter" data-level="20.1" data-path="overdispersion.html"><a href="overdispersion.html#recognising-and-testing-for-overdispersion"><i class="fa fa-check"></i><b>20.1</b> Recognising (and testing for) overdispersion</a></li>
<li class="chapter" data-level="20.2" data-path="overdispersion.html"><a href="overdispersion.html#refefernce"><i class="fa fa-check"></i><b>20.2</b> Refefernce</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="package-management.html"><a href="package-management.html"><i class="fa fa-check"></i><b>21</b> Package management</a></li>
<li class="chapter" data-level="22" data-path="permutation-and-combination.html"><a href="permutation-and-combination.html"><i class="fa fa-check"></i><b>22</b> Permutation and combination</a><ul>
<li class="chapter" data-level="22.1" data-path="permutation-and-combination.html"><a href="permutation-and-combination.html#reference-17"><i class="fa fa-check"></i><b>22.1</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="23" data-path="partial-least-square-analysis.html"><a href="partial-least-square-analysis.html"><i class="fa fa-check"></i><b>23</b> Partial least square analysis</a><ul>
<li class="chapter" data-level="23.1" data-path="partial-least-square-analysis.html"><a href="partial-least-square-analysis.html#prepare-the-data-1"><i class="fa fa-check"></i><b>23.1</b> Prepare the data</a></li>
<li class="chapter" data-level="23.2" data-path="partial-least-square-analysis.html"><a href="partial-least-square-analysis.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>23.2</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="23.3" data-path="partial-least-square-analysis.html"><a href="partial-least-square-analysis.html#partial-least-squares-pls-and-pls-da"><i class="fa fa-check"></i><b>23.3</b> Partial least-squares: PLS and PLS-DA</a></li>
<li class="chapter" data-level="23.4" data-path="partial-least-square-analysis.html"><a href="partial-least-square-analysis.html#reference-18"><i class="fa fa-check"></i><b>23.4</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="scale.html"><a href="scale.html"><i class="fa fa-check"></i><b>24</b> Scale</a><ul>
<li class="chapter" data-level="24.1" data-path="scale.html"><a href="scale.html#reference-19"><i class="fa fa-check"></i><b>24.1</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="web-scraping.html"><a href="web-scraping.html"><i class="fa fa-check"></i><b>25</b> Web scraping</a><ul>
<li class="chapter" data-level="25.1" data-path="web-scraping.html"><a href="web-scraping.html#download-content-from-ncbi-databases-using-rismed"><i class="fa fa-check"></i><b>25.1</b> Download Content from NCBI Databases using <code>RISmed</code></a><ul>
<li class="chapter" data-level="25.1.1" data-path="web-scraping.html"><a href="web-scraping.html#search-author-name"><i class="fa fa-check"></i><b>25.1.1</b> Search author name</a></li>
<li class="chapter" data-level="25.1.2" data-path="web-scraping.html"><a href="web-scraping.html#search-keyword"><i class="fa fa-check"></i><b>25.1.2</b> Search keyword</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="26" data-path="linear-mixed-model-in-r.html"><a href="linear-mixed-model-in-r.html"><i class="fa fa-check"></i><b>26</b> Linear mixed model in R</a><ul>
<li class="chapter" data-level="26.1" data-path="linear-mixed-model-in-r.html"><a href="linear-mixed-model-in-r.html#data-from-an-oats-field-trial"><i class="fa fa-check"></i><b>26.1</b> Data from an Oats Field Trial</a></li>
<li class="chapter" data-level="26.2" data-path="linear-mixed-model-in-r.html"><a href="linear-mixed-model-in-r.html#data-visulization"><i class="fa fa-check"></i><b>26.2</b> Data visulization</a></li>
<li class="chapter" data-level="26.3" data-path="linear-mixed-model-in-r.html"><a href="linear-mixed-model-in-r.html#mixed-linear-model"><i class="fa fa-check"></i><b>26.3</b> Mixed linear model</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Notes of R for Bioinformatics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="correlation" class="section level1">
<h1><span class="header-section-number">3</span> Correlation</h1>
<div id="visulization-of-pair-wise-correlation-in-r" class="section level2">
<h2><span class="header-section-number">3.1</span> Visulization of pair-wise correlation in R</h2>
</div>
<div id="correlation-and-p-values-of-all-combinations-of-all-rows-of-two-matrices" class="section level2">
<h2><span class="header-section-number">3.2</span> Correlation and p-values of all combinations of all rows of two matrices</h2>
<div id="prepare-the-data" class="section level3">
<h3><span class="header-section-number">3.2.1</span> Prepare the data</h3>
<pre class="sourceCode r"><code class="sourceCode r">bac1 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>)
bac2 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">1</span>)
bac3 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>)
bac4 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>)
bac &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">cbind</span>(bac1, bac2, bac3, bac4 ))
<span class="kw">colnames</span>(bac) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;station1&quot;</span>, <span class="st">&quot;station2&quot;</span>, <span class="st">&quot;station3&quot;</span>, <span class="st">&quot;station4&quot;</span>)
<span class="kw">rownames</span>(bac) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;bac1&quot;</span>, <span class="st">&quot;bac2&quot;</span>, <span class="st">&quot;bac3&quot;</span>, <span class="st">&quot;bac4&quot;</span>, <span class="st">&quot;bac5&quot;</span>)

fac1 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>)
fac2 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">1</span>,<span class="dv">6</span>)
fac3&lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>)
fac4&lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>, <span class="dv">6</span>)
fac &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">cbind</span>(fac1, fac2, fac3, fac4))
<span class="kw">colnames</span>(fac) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;station1&quot;</span>, <span class="st">&quot;station2&quot;</span>, <span class="st">&quot;station3&quot;</span>, <span class="st">&quot;station4&quot;</span>)
<span class="kw">rownames</span>(fac) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;fac1&quot;</span>, <span class="st">&quot;fac2&quot;</span>, <span class="st">&quot;fac3&quot;</span>, <span class="st">&quot;fac4&quot;</span>, <span class="st">&quot;fac5&quot;</span>, <span class="st">&quot;fac6&quot;</span>)

bac</code></pre>
<pre><code>##      station1 station2 station3 station4
## bac1        1        2        4        5
## bac2        2        3        5        1
## bac3        3        4        1        2
## bac4        4        5        2        3
## bac5        5        1        3        4</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">fac</code></pre>
<pre><code>##      station1 station2 station3 station4
## fac1        1        2        3        4
## fac2        2        3        4        5
## fac3        3        4        5        1
## fac4        4        5        1        2
## fac5        5        1        2        3
## fac6        6        6        1        6</code></pre>
</div>
<div id="another-way-to-do-this-is-to-use-corr.test" class="section level3">
<h3><span class="header-section-number">3.2.2</span> Another way to do this is to use <code>corr.test</code></h3>
<p>You can just pass the full matrices to the <code>cor</code> function (or psych::corr.test)and it takes care of finding the correlation of the relevant columns.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">#install.packages(&quot;psych&quot;)</span>
<span class="kw">library</span>(reshape2)
<span class="co">#cor.eff&lt;-cor(t(fac), t(bac))</span>
<span class="co">#coreff &lt;- reshape2::melt(cor.eff)</span>
pval &lt;-<span class="st"> </span>psych<span class="op">::</span><span class="kw">corr.test</span>(<span class="kw">t</span>(fac), <span class="kw">t</span>(bac), <span class="dt">adjust=</span><span class="st">&quot;fdr&quot;</span>)
pval</code></pre>
<pre><code>## Call:psych::corr.test(x = t(fac), y = t(bac), adjust = &quot;fdr&quot;)
## Correlation matrix 
##       bac1  bac2  bac3  bac4  bac5
## fac1  0.99 -0.08 -0.60 -0.60 -0.08
## fac2  0.99 -0.08 -0.60 -0.60 -0.08
## fac3 -0.32  0.94 -0.08 -0.08 -0.49
## fac4 -0.80 -0.32  0.99  0.99 -0.32
## fac5 -0.32 -0.49 -0.08 -0.08  0.94
## fac6 -0.37 -0.88  0.77  0.77  0.10
## Sample Size 
## [1] 4
## Probability values  adjusted for multiple tests. 
##      bac1 bac2 bac3 bac4 bac5
## fac1 0.08 0.92 0.86 0.86 0.92
## fac2 0.08 0.92 0.86 0.86 0.92
## fac3 0.92 0.29 0.92 0.92 0.92
## fac4 0.68 0.92 0.08 0.08 0.92
## fac5 0.92 0.92 0.92 0.92 0.29
## fac6 0.92 0.52 0.68 0.68 0.92
## 
##  To see confidence intervals of the correlations, print with the short=FALSE option</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">cor_pval&lt;-<span class="st"> </span><span class="kw">merge</span>(<span class="kw">melt</span>(pval<span class="op">$</span>r, <span class="dt">value.name=</span><span class="st">&quot;cor&quot;</span>), <span class="kw">melt</span>(pval<span class="op">$</span>p, <span class="dt">value.name=</span><span class="st">&quot;p-value&quot;</span>), <span class="dt">by=</span><span class="kw">c</span>(<span class="st">&quot;Var1&quot;</span>, <span class="st">&quot;Var2&quot;</span>))
<span class="kw">head</span>(cor_pval)</code></pre>
<pre><code>##   Var1 Var2         cor   p-value
## 1 fac1 bac1  0.98994949 0.0753788
## 2 fac1 bac2 -0.07559289 0.9244071
## 3 fac1 bac3 -0.60000000 0.8571429
## 4 fac1 bac4 -0.60000000 0.8571429
## 5 fac1 bac5 -0.07559289 0.9244071
## 6 fac2 bac1  0.98994949 0.0753788</code></pre>
</div>
<div id="read-data" class="section level3">
<h3><span class="header-section-number">3.2.3</span> Read data</h3>
<pre><code>data/cor_data_metablomics.txt         
data/cor_data_transcriptome.txt
</code></pre>
</div>
<div id="calculate-the-correlation-advance" class="section level3">
<h3><span class="header-section-number">3.2.4</span> Calculate the correlation (Advance)</h3>
<pre class="sourceCode r"><code class="sourceCode r">tbac &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">t</span>(bac))
tfac &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">t</span>(fac))

f &lt;-<span class="st"> </span><span class="cf">function</span> (x, y) <span class="kw">cor</span>(x, y)
pval &lt;-<span class="st"> </span><span class="cf">function</span>(x, y) <span class="kw">cor.test</span>(x, y)<span class="op">$</span>p.val

tab &lt;-<span class="st"> </span><span class="kw">outer</span>(tfac, tbac, <span class="kw">Vectorize</span>(f))
tabp&lt;-<span class="kw">outer</span>(tfac, tbac, <span class="kw">Vectorize</span>(pval))

<span class="kw">as.data.frame.table</span>(tabp)</code></pre>
<pre><code>##    Var1 Var2       Freq
## 1  fac1 bac1 0.01005051
## 2  fac2 bac1 0.01005051
## 3  fac3 bac1 0.67928651
## 4  fac4 bac1 0.20000000
## 5  fac5 bac1 0.67928651
## 6  fac6 bac1 0.63485163
## 7  fac1 bac2 0.92440711
## 8  fac2 bac2 0.92440711
## 9  fac3 bac2 0.05714286
## 10 fac4 bac2 0.67928651
## 11 fac5 bac2 0.51428571
## 12 fac6 bac2 0.12168993
## 13 fac1 bac3 0.40000000
## 14 fac2 bac3 0.40000000
## 15 fac3 bac3 0.92440711
## 16 fac4 bac3 0.01005051
## 17 fac5 bac3 0.92440711
## 18 fac6 bac3 0.22540333
## 19 fac1 bac4 0.40000000
## 20 fac2 bac4 0.40000000
## 21 fac3 bac4 0.92440711
## 22 fac4 bac4 0.01005051
## 23 fac5 bac4 0.92440711
## 24 fac6 bac4 0.22540333
## 25 fac1 bac5 0.92440711
## 26 fac2 bac5 0.92440711
## 27 fac3 bac5 0.51428571
## 28 fac4 bac5 0.67928651
## 29 fac5 bac5 0.05714286
## 30 fac6 bac5 0.90240999</code></pre>
</div>
<div id="reference-3" class="section level3">
<h3><span class="header-section-number">3.2.5</span> Reference</h3>
<p>Correlation/p values of all combinations of all rows of two matrices: <a href="https://stackoverflow.com/questions/41793219/correlation-p-values-of-all-combinations-of-all-rows-of-two-matrices" class="uri">https://stackoverflow.com/questions/41793219/correlation-p-values-of-all-combinations-of-all-rows-of-two-matrices</a></p>
</div>
</div>
<div id="different-ways-to-visulize-correlation" class="section level2">
<h2><span class="header-section-number">3.3</span> Different ways to visulize correlation</h2>
<div id="display-correlation-using-chart.correlation-in-performanceanalytics" class="section level3">
<h3><span class="header-section-number">3.3.1</span> Display correlation using <code>chart.Correlation</code> in <code>PerformanceAnalytics</code></h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(datasets)
<span class="kw">data</span>(iris)
<span class="kw">library</span>(<span class="st">&quot;PerformanceAnalytics&quot;</span>)</code></pre>
<pre><code>## Loading required package: xts</code></pre>
<pre><code>## Loading required package: zoo</code></pre>
<pre><code>## 
## Attaching package: &#39;zoo&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     as.Date, as.Date.numeric</code></pre>
<pre><code>## 
## Attaching package: &#39;PerformanceAnalytics&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:graphics&#39;:
## 
##     legend</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">chart.Correlation</span>(iris[, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>], <span class="dt">histogram=</span><span class="ot">TRUE</span>, <span class="dt">pch=</span><span class="dv">19</span>)</code></pre>
<p><img src="bioinfBookXIE186_files/figure-html/unnamed-chunk-13-1.png" width="576" /></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(corrplot)</code></pre>
<pre><code>## corrplot 0.84 loaded</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">corrplot.mixed</span>(<span class="kw">cor</span>(iris[, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>]), <span class="dt">order=</span><span class="st">&quot;original&quot;</span>, <span class="dt">tl.col=</span><span class="st">&quot;black&quot;</span>)</code></pre>
<p><img src="bioinfBookXIE186_files/figure-html/unnamed-chunk-14-1.png" width="576" /></p>
</div>
</div>
<div id="reference-4" class="section level2">
<h2><span class="header-section-number">3.4</span> Reference</h2>
<p>Seven Easy Graphs to Visualize Correlation Matrices in R: <a href="http://jamesmarquezportfolio.com/correlation_matrices_in_r.html" class="uri">http://jamesmarquezportfolio.com/correlation_matrices_in_r.html</a></p>
<p>Example 9.17: (much) better pairs plots: <a href="https://www.r-bloggers.com/example-9-17-much-better-pairs-plots/" class="uri">https://www.r-bloggers.com/example-9-17-much-better-pairs-plots/</a></p>
<p>Introduction to Correlation: <a href="https://rpubs.com/aaronsc32/linear-relationship-pearson-r-correlation" class="uri">https://rpubs.com/aaronsc32/linear-relationship-pearson-r-correlation</a></p>
<p>Spearman Rank Correlation: <a href="https://rstudio-pubs-static.s3.amazonaws.com/191093_4169c5282eb145a491a5b1924941a6ba.html" class="uri">https://rstudio-pubs-static.s3.amazonaws.com/191093_4169c5282eb145a491a5b1924941a6ba.html</a></p>

</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="plot-in-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="standard-deviation-vs-standard-error.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
